# Lab 02 - Bài tập

---

## THÔNG TIN

- Họ tên: Nguyen Truong Phat
- MSSV: 23697331
- Ngày hoàn thành:

---
## NUMPY

Bạn hãy phát sinh ngẫu nhiên mảng ndarray có tên `arr` gồm 100 phần tử số nguyên có giá trị từ 0..1000. Thực hiện các thao tác sau:

### 1. Hiển thị các phần tử thỏa yêu cầu dưới đây

- 10 phần tử đầu tiên
- 5 phần tử cuối cùng
- Các phần tử thỏa mãn điều kiện chia hết cho 5 và lớn hơn 100

In [ ]:
import numpy as np
from numpy import random
a = random.randint(0, 1000, size=(100))
print(a)
print("10 phần tử đầu tiên" ,a[0:10])
print("5 phần tử cuối cùng",a[-5:])
print("Các phần tử thỏa mãn điều kiện chia hết cho 5 và lớn hơn 100",a[(a % 5 == 0) & (a > 100)])

[902 191   6 729 133 414  97 951 725 794 983 717 707  95 765 403 906  72
 658 398 739 933 407 929 447 582  71 158 210 624 261 993 663 792 750  11
 209   5  64 711  21 913 943 694 970 187 632 414 723 744 516 907  78 428
 417 797 822 192 113 148 900  65 478 345 659 464 402 171 222 345 389 124
 364 333 872 749 485 976 237  80 551 178 788 262 317 729 338 595 335 886
 854  96 883 957 804 430 471  55 774 627]
10 phần tử đầu tiên [902 191   6 729 133 414  97 951 725 794]
5 phần tử cuối cùng [430 471  55 774 627]
Các phần tử thỏa mãn điều kiện chia hết cho 5 và lớn hơn 100 [725 765 210 750 970 900 345 345 485 595 335 430]


### 2. Cho biết mảng trên có bao nhiêu phần tử có giá trị nhỏ hơn 100. Hiển thị các phần tử đó

In [ ]:
len(a[a<100])

14

### 3. Thực hiện thống kê:

- Tính tổng mảng
- Tìm giá trị trung bình
- Tìm giá trị trung vị
- Tìm Q1, Q2, Q3
- Tìm phân vị 20 (percentile 20)

In [ ]:
print("Tổng mảng",a.sum())
print("Giá trị trung bình",a.mean())
print("Giá trị trung vị",np.median(a))
print('Tứ phân vị [Q1, Q2, Q3]: ', np.quantile(a, [0.25, 0.5, 0.75]))
print('Phân vị 20 (median): ', np.quantile(a,0.2))

Tổng mảng 50454
Giá trị trung bình 504.54
Giá trị trung vị 474.5
Tứ phân vị [Q1, Q2, Q3]:  [209.75 474.5  767.25]
Phân vị 20 (median):  176.6


---

## PANDAS

Thông tin các bộ dữ liệu:

- `car-sales.csv`: bộ dữ liệu chứa thông tin của các ôtô cũ ở một cửa hàng


### 1. Đọc dữ liệu

In [21]:
import pandas as pd
import numpy as np

car = pd.read_csv('car-sales.csv')
car

Manufacturer Colour  Odometer  Doors    Price
0       Toyota  White    150043      4   4000.0
1        Honda    Red     87899      4   5000.0
2       Toyota   Blue     32549      3   7000.0
3          BMW  Black     11179      5  22000.0
4       Nissan  White    213095      4   3500.0
5       Toyota  Green     99213      4   4500.0
6        Honda   Blue     45698      4   7500.0
7        Honda   Blue     54738      4   7000.0
8       Toyota  White     60000      4   6250.0
9       Nissan  White     31600      4   9700.0

### 2. Dữ liệu có bao nhiêu dòng, bao nhiêu cột

Đáp án: (10 dòng, 5 cột)

In [4]:
car.shape

(10, 5)

- Định tính (phân loại): Manufacturer, Colour
- Định lượng: Odemeter, Price, Doors

In [5]:
car.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10 entries, 0 to 9
Data columns (total 5 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Manufacturer  10 non-null     object 
 1   Colour        10 non-null     object 
 2   Odometer      10 non-null     int64  
 3   Doors         10 non-null     int64  
 4   Price         10 non-null     float64
dtypes: float64(1), int64(2), object(2)
memory usage: 528.0+ bytes


### 3. Trong dữ liệu có thông tin của bao nhiêu xe?

Đáp án: 10 xe

In [11]:
data = car.Manufacturer.count()
data

10

### 4. Đếm số xe theo theo hãng

Đáp án: Toyota: 4 xe, Honda: 3 xe, Nissan: 2 xe, BMW: 1 xe

In [12]:
soxe = car.Manufacturer.value_counts()
soxe

Manufacturer
Toyota    4
Honda     3
Nissan    2
BMW       1
Name: count, dtype: int64

### 5. Đếm số xe theo màu


In [13]:
color = car.Colour.value_counts()
color

Colour
White    4
Blue     3
Red      1
Black    1
Green    1
Name: count, dtype: int64

### 6. Tổng giá trị các xe trong cửa hàng

Đáp án: 76450 USD


In [14]:
sumPrice = car.Price.sum()
sumPrice

76450.0

### 7. Top 3 xe có giá cao nhất


In [15]:
maxPrice = car.nlargest(3, 'Price')
maxPrice

Manufacturer Colour  Odometer  Doors    Price
3          BMW  Black     11179      5  22000.0
9       Nissan  White     31600      4   9700.0
6        Honda   Blue     45698      4   7500.0

### 8. Top 3 xe đã chạy nhiều cây số (km)nhất (Odometer)

Đáp án:
- Xe Nissan, White, 213095 km
- Xe Toyota, White, 150043 km
- Xe Toyota, Green, 99213 km

In [17]:
Odometer = car.nlargest(3, 'Odometer')
Odometer

Manufacturer Colour  Odometer  Doors   Price
4       Nissan  White    213095      4  3500.0
0       Toyota  White    150043      4  4000.0
5       Toyota  Green     99213      4  4500.0

### 9. Khảo sát xe của hãng Honda

- Tính tỷ lệ xe Honda trong dữ liệu của cửa hàng (0.3)
- Cho biết giá trung bình của xe hãng Honda của cửa hàng (6500)
- Cho biết các xe của hãng Honda trong cửa hàng có những màu sắc nào (Red, Blue)

In [25]:
honda_cars_count = car['Manufacturer'].value_counts()['Honda']
print('Tỷ lệ xe Honda trong dữ liệu của cửa hàng',honda_cars_count/car.shape[0])
honda_cars_avg_price = car[car['Manufacturer'] == 'Honda']['Price'].mean()
print('Giá trung bình của xe hãng Honda của cửa hàng',honda_cars_avg_price)
honda_cars_color = car[car['Manufacturer'] == 'Honda']['Colour'].unique()
print('Xe của hãng Honda trong cửa hàng có những màu sắc nào',honda_cars_color)


Tỷ lệ xe Honda trong dữ liệu của cửa hàng 0.3
Giá trung bình của xe hãng Honda của cửa hàng 6500.0
Xe của hãng Honda trong cửa hàng có những màu sắc nào ['Red' 'Blue']


### 10. Tính toán các giá trị sau của  thuộc tính giá ('Price')

- Giá trung bình các xe (7645)
- Giá trị trung vị các xe (6625)
- Miền giá trị của xe (18500)
- Miền phân vị của xe (2750)
- Điểm phân vị 0.4 và 0.6 của giá ([5750, 7000])

In [36]:
print('Giá trung bình các xe',car['Price'].mean())
print('Giá trị trung vị các xe',car['Price'].median())
print('Miền giá trị của xe',car['Price'].max()-car['Price'].min())
print('Miền phân vị của xe',car['Price'].quantile(0.75) -car['Price'].quantile(0.25))
print('Điểm phân vị 0.4 và 0.6 của giá',car['Price'].quantile([0.4,0.6]))

Giá trung bình các xe 7645.0
Giá trị trung vị các xe 6625.0
Miền giá trị của xe 18500.0
Miền phân vị của xe 2750.0
Điểm phân vị 0.4 và 0.6 của giá 0.4    5750.0
0.6    7000.0
Name: Price, dtype: float64


---